crates/matrix-sdk/tests/integration/encryption/backups.rs

```diff
@@ -1361,6 +1372,131 @@ async fn enable_from_secret_storage_and_download_after_utd() {
+    let room_id = room_id!("!DovneieKSTkdHKpIXy:morpheus.localhost");
+    let event_id = event_id!("$JbFHtZpEJiH8uaajZjPLz0QUZc1xtBR9rPGBOjF6WFM");
+
+    let session = MatrixSession {
+        meta: SessionMeta { user_id: user_id.into(), device_id: device_id!("DEVICEID").to_owned() },
+        tokens: MatrixSessionTokens { access_token: "1234".to_owned(), refresh_token: None },
+    };
+    let (builder, server) = test_client_builder_with_server().await;
+    let encryption_settings = EncryptionSettings {
+        backup_download_strategy: BackupDownloadStrategy::AfterDecryptionFailure,
+        ..Default::default()
+    };
+    let client = builder
+        .request_config(RequestConfig::new().disable_retry())
+        .with_encryption_settings(encryption_settings)
+        .build()
+        .await
+        .unwrap();
+
+    client.restore_session(session).await.unwrap();
+
+    let sync = SyncResponseBuilder::new()
+        .add_joined_room(JoinedRoomBuilder::new(room_id))
+        .build_json_sync_response();
+    mock_sync(&server, sync, None).await;
+
+    client.sync_once(Default::default()).await.expect("We should be able to sync with the server");
+
+    init_client_secret_storage_and_backup(&client, &server).await;
+
+    // Create an outbound group session which we will use to encrypt a test event.
+    let sender_identity_keys = IdentityKeys {
+        ed25519: Ed25519SecretKey::new().public_key(),
+        curve25519: Curve25519PublicKey::from(&Curve25519SecretKey::new()),
+    };
+    let outbound_group_session = OutboundGroupSession::new(
+        device_id!("KIUVQQSDTM").to_owned(),
+        Arc::new(sender_identity_keys),
+        room_id,
+        matrix_sdk::crypto::EncryptionSettings::default(),
+    )
+    .unwrap();
+
+    // Export the `OutboundGroupSession` to an `InboundGroupSession`, and export it
+    // to the backup. We do this now, at ratchet index 0.
+    let inbound_group_session = inbound_session_from_outbound_session(
+        sender_identity_keys.ed25519,
+        room_id,
+        &outbound_group_session,
+    )
+    .await
+    .unwrap();
+    mock_download_session_from_key_backup(room_id, inbound_group_session, &server).await;
+
+    // Encrypt an event and prepare for the client to download it.
+    let event_body = json!({"body":"tt","msgtype":"m.text"});
+    let encrypted_event_content = serde_json::to_value(
+        outbound_group_session
+            .encrypt("m.room.message", &serde_json::from_value(event_body).unwrap())
+            .await,
+    )
+    .unwrap();
+    mock_get_event(room_id, event_id, encrypted_event_content, &server).await;
+
+    // Now, import the megolm session into the client's store, at ratchet index 1.
+    {
+        let inbound_group_session = inbound_session_from_outbound_session(
+            sender_identity_keys.ed25519,
+            room_id,
+            &outbound_group_session,
+        )
+        .await
+        .unwrap();
```

Comment:
Maybe an assert that checks that the message index is indeed `1` wouldn't hurt here.